cpp/include/cudf/detail/offsets_iterator_factory.cuh

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/offsets_iterator.cuh>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Use this class to create an offsetalator instance.
+ */
+struct offsetalator_factory {
+  /**
+   * @brief A type_dispatcher functor to create an input iterator from an offsets column
+   */
+  struct input_offsetalator_fn {
+    template <typename T, CUDF_ENABLE_IF(std::is_same_v<T, int32_t> or std::is_same_v<T, int64_t>)>
+    input_offsetalator operator()(column_view const& indices)
```

Comment:
`input_offselator` already has a runtime guard against `indices.type()` not containing values either `INT32` or `INT64`. This type-dispatcher functor (and by extension the type-dispatcher of the factory) can be entirely removed, and you can forward the pointer as `indices.data<void>()` or `indices.data<char>()`